src/interpreter/InterpreterRelation.h

```diff
@@ -31,244 +31,405 @@
+        return tuple;
     }
 
     /**
-     * Add all entries of the given relation to this relation.
+     * Cast an abstract view into a view of Index::InterpreterView type.
      */
-    void insert(const InterpreterRelation& other);
+    static View* castView(InterpreterViewWrapper* view) {
+        return static_cast<View*>(view);
+    }
 
     /**
-     * Tests whether this relation contains the given tuple.
+     * Creates a relation, build all necessary indexes.
      */
-    bool contains(const TupleRef& tuple) const;
+    InterpreterRelation(
+            std::size_t auxiliaryArity, std::string name, const ram::analysis::MinIndexSelection& orderSet)
+            : InterpreterRelationWrapper(Arity, auxiliaryArity, std::move(name)) {
+        for (auto order : orderSet.getAllOrders()) {
+            // Expand the order to a total order
+            ram::analysis::MinIndexSelection::AttributeSet set{order.begin(), order.end()};
+
+            // This operation is not performance critical.
+            // Not using constexpr Arity to avoid compiler warning. (When Arity == 0)
+            for (std::size_t i = 0; i < getArity(); ++i) {
+                if (set.find(i) == set.end()) {
+                    order.push_back(i);
+                }
+            }
+
+            indexes.push_back(mk<Index>(order));
+        }
 
-    /**
-     * Tests whether this relation contains any element between the given boundaries.
-     */
-    bool contains(const size_t& indexPos, const TupleRef& low, const TupleRef& high) const;
+        // Use the first index as default main index
+        main = indexes[0].get();
+    }
 
-    /**
-     * Obtains a stream to scan the entire relation.
-     */
-    Stream scan() const;
+    InterpreterRelation(InterpreterRelation& other) = delete;
 
-    /**
-     * Obtains a partitioned stream list for parallel computation
-     */
-    PartitionedStream partitionScan(size_t partitionCount) const;
+    // -- Implement all virtual interface from Wrapper. --
+    // -- Operations defined in this section are not performance-oriented.
+public:
+    void purge() override {
+        __purge();
+    }
+
+    void insert(const RamDomain* data) override {
+        insert(constructTuple(data));
+    }
 
+    bool contains(const RamDomain* data) const override {
+        return contains(constructTuple(data));
+    }
+
+    IndexViewPtr createView(const size_t& indexPos) const override {
+        return mk<View>(indexes[indexPos]->createView());
+    }
+
+    size_t size() const override {
+        return __size();
+    }
+
+    Order getIndexOrder(size_t idx) const override {
+        return indexes[idx]->getOrder();
+    }
+
+    class iterator_base : public InterpreterRelationWrapper::iterator_base {
+        iterator iter;
+        Order order;
+        RamDomain data[Arity];
+
+    public:
+        iterator_base(const typename Index::iterator& iter, Order order)
+                : iter(iter), order(std::move(order)) {}
```

Comment:
One tiny change more:
```suggestion
        iterator_base(typename Index::iterator iter, Order order)
                : iter(std::move(iter)), order(std::move(order)) {}
```
If we're going to make a copy anyway, then don't take a const ref.

Reply:
Yeah, good point.